src/cdk/tree/navigator.ts

```diff
@@ -0,0 +1,311 @@
+export interface CdkNavigator<T> {
+  insert(index: number, data: T, node: FocusableOption, parentData?: T);
+  remove(index: number, parentData?: any);
+  move(previousIndex: number, currentIndex: number, parentData?: T);
+  updateFocusedData(newFocusedData: T);
+}
+
+/**
+ * Navigator for CDK tree component. Use keyboard to navigate through the tree nodes.
+ * `HOME` moves the focus to first node in the tree
+ * `END` moves the focus to last node in the tree
+ * `UP_ARROW`, and `DOWN_ARROW` moves the focus to previous/next visible node.
+ * `LEFT_ARROW` moves the focus to parent node, or collapse current focused node.
+ * `RIGHT_ARROW` moves the focus to first child node, or expand the current focused node.
+ */
+@Directive({
+  selector: '[cdkTreeNavigator]',
+  host: {
+    '(keydown)': '_handleKeydown($event)',
+    'class': 'cdk-tree-navigator',
+  }
+})
+export class CdkTreeNavigator<T> implements CdkNavigator<T> {
+  /** The node map map data nodes to CdkTreeNodes */
+  protected nodeMap: Map<T, FocusableOption> = new Map<T, FocusableOption>();
+
+  /** A map from parent node to a list of children. */
+  protected nodeListMap: Map<T | undefined, T[]> = new Map<T | undefined, T[]>();
+
+  /** A map from node data to its parent node data. */
+  protected parentMap: Map<T, T> = new Map<T, T>();
+
+  /** Current focused node data. */
+  focusedData: T;
+
+  /** Tree control is used to expand or collapse nodes. */
+  treeControl: TreeControl<T>;
+
+  /**
+   * Stream that emits any time the TAB key is pressed, so components can react
+   * when focus is shifted off of the list.
+   */
+  tabOut: Subject<void> = new Subject<void>();
+
+  constructor(@Optional() protected dir: Directionality) {}
+
+  /** Add tree node to navigator's data */
+  insert(index: number, data: T, node: FocusableOption, parentData?: T) {
+    this.nodeMap.set(data, node);
+    if (parentData) {
+      this.parentMap.set(data, parentData);
+    }
+    this._getNodeList(parentData).splice(index, 0, data);
+  }
+
+  /** Remove a node data from navigator */
+  remove(index: number, parentData?: T) {
+    const nodeList = this._getNodeList(parentData);
+    const removed = nodeList.splice(index, 1)[0];
+    if (removed && removed === this.focusedData) {
+      this._changeFocusedData(this._getNext());
+    }
+    this.nodeMap.delete(removed);
+    this.parentMap.delete(removed);
+    this.nodeListMap.delete(removed);
+  }
+
+  /** Update node's index information */
+  move(previousIndex: number, currentIndex: number, parentData?: T) {
+    const nodeList = this._getNodeList(parentData);
+    const target = nodeList.splice(previousIndex, 1);
+    nodeList.splice(currentIndex, 0, target[0]);
+  }
+
+  /** When a tree node is focused, update the focused data in navigator */
+  updateFocusedData(newFocusedData: T) {
+    this.focusedData = newFocusedData;
+  }
+
+  /** Change focus to first visible node in the tree. */
+  focusFirst() {
```

Comment:
Were you able to test this out in VoiceOver and/or NVDA to see how it did? I'm wondering how using focus and aria-activedescendant compare

Reply:
Tested in VoiceOver for `focus` and `aria-activedescendant`. The reader reads the same thing for both, but in `aria-activedescendant` version, the first node is not actually focused. The focus stays on the whole tree.

Reply:
Isn't that the way `aria-activedescendant` is supposed to work?